### DISCUSSION TITLE
Logs: adapt to v6 before stable version is out

### DISCUSSION THREAD

**Discussion by ZhengRui** [fabricjs/fabric.js#8739] [Q&A] [answered]
I am trying to adapt my code to v6. I see `canvas.getObjects()` now returns `BaseFabricObject`, in v5, it returns `fabric.Object`, and all my code was based on `fabric.Object`. I notice there is still `fabric.Object` in v6,  so I am wondering what are their differences? which one should I use in general? should I do something like `canvas.getObjects() as fabric.Object[]` ?

**Comment by ZhengRui** [comment]
A follow up question: I found `setOptions` is not exposed in v6, in v5 I used to use `set()` for default props and `setOptions()` to bind data with objects. Should I use `rect.prop1 = val1; rect.prop2 = val2` or `set({prop1: val1, prop2: val2})` to set custom props (or in my words bind data) now?

**Comment by ShaMan123** [comment, marked as answer]
`BaseFabricObject` - not interactive, just for rendering, supposed to be used with StaticCanvas, for apps that want to produce visuals w/o interaction, probably node apps, supposed to reduce bundle size eventually but we haven't exposed a way of using it yet because all shapes should subclass it as well (along side `FabricObject`) to match the concept so for now it is useless.
`FabricObject === fabric.Object`

In canvas it is a type issue that needs fixing

`setOptions` => `set`

**Comment by ZhengRui** [comment]
Log of my steps to adapt my code from v5.3 to v6.0.0-beta1, hope it can be helpful.

1. `import { fabric } from 'fabric'` -> `import * as fabric from 'fabric'`
2. There are some types that are not exposed yet, e.g. TMat2D, events. Before v6 stable comes out, you can expose it by add 
```json
"src/*": [
  "dist/src/*"
]
```
to the `typesVersions` section. Note that it should be `dist/src/*` instead of `src/*` as mentioned here [pull/8725](https://github.com/fabricjs/fabric.js/pull/8725/files), [discussions/8715](https://github.com/fabricjs/fabric.js/discussions/8715), [pull/8717](https://github.com/fabricjs/fabric.js/pull/8717), otherwise it has issue (relates to event types) saying event typed in `src/**` can not be assigned to the same event typed in `dist/src/**`. Use `dist/src/*` will ensure all `fabric` objects and types are from `dist` therefore consistent.
   - `import { TMat2D } from 'fabric/src/typedefs';` TMat2D is required for `canvas.setViewportTransform()`
   - for events, v5's `fabric.IEvent` -> `import { TPointerEventInfo as IEvent } from 'fabric/src/EventTypeDefs';`
3. `canvas.getObjects()` now returns `BaseFabricObject` ([discussion/8739](https://github.com/fabricjs/fabric.js/discussions/8739#discussioncomment-5118207)), you have to cast to `fabric.Object` if it is expected in downstream logic 
4. `setOptions()` not needed anymore, just use `set()` for all props including custom props
5. `setCoords()` not needed anymore, just delete it, seems v6 handles that automatically now (not sure though)
6. `new fabric.Circle({...props})` has types issue, it asks for full `CircleProps` which has over 50 props to fill. You can hack it by 
```typescript
const circle = new fabric.Circle()  // this is fine
circle.set({...props})
```
7. v5 event's `pointer` used to be absolute coordinates, and `setViewportTransform()` expects absolute coordinates. In v6 event's `pointer` is canvas coordinates, but it returns `absolutePointer` which gives absolute coordinates needed for viewport transform. 
8. In v6, `canvas.remove()` will trigger `object:modified` event if the removed object is also active. In v5 it does not trigger `object:modified` event.
```typescript
'object:modified': (e: IEvent<MouseEvent>) => {
  if (!e.e) return; // likely triggered by canvas.remove()

  // actually modified
},
``` 
9. `canvas.setBackgroundImage(imageObj, () => {})` -> `canvas.backgroundImage = imageObj`
10. 
```typescript
const idx = canvas.getObjects().indexOf(line);
canvas.remove(line).add(line_);
line_.moveTo(idx)
``` 
-> 
```typescript
const idx = canvas.getObjects().indexOf(line);
canvas.remove(line);
canvas.add(line_);
canvas.moveObjectTo(line_, idx);
```

**Reply by ShaMan123** [reply]
This is great!

---


> 5\. `setCoords()` not needed anymore, just delete it, seems v6 handles that automatically now (not sure though)

Not yet, but it might be #8682 



> 6\. `new fabric.Circle({...props})` has types issue, it asks for full `CircleProps` which has over 50 props to fill. You can hack it by

#8677 



> 7\. v5 event's `pointer` used to be absolute coordinates, and `setViewportTransform()` expects absolute coordinates. In v6 event's `pointer` is canvas coordinates, but it returns `absolutePointer` which gives absolute coordinates needed for viewport transform.

I don't think so, please elaborate with a repro



> 8\. In v6, `canvas.remove()` will trigger `object:modified` event if the removed object is also active. In v5 it does not trigger `object:modified` event.

Shouldn't happen, please provide a repro

**Reply by ZhengRui** [reply]
Btw my above conclusions was based on [master@196bea17](https://github.com/fabricjs/fabric.js/tree/196bea176c4b01fd68a05aa096ba5d22459201b8)

7.
 - in v5, i have code like this to implement zooming/panning of canvas:
```typescript
function parseEvent<T extends MouseEvent | WheelEvent>(
  e: fabric.IEvent<T>
) {
  const { button, target, pointer, e: evt } = e;
  evt.preventDefault();
  evt.stopPropagation();

  return { button, target, pointer, evt };
}

const setZoomByWheel = (evt: WheelEvent) => {
  const { deltaY, offsetX: x, offsetY: y } = evt;

  // make touchBoard more smooth, using ctrlKey to identify touchBoard
  // more detail in: https://use-gesture.netlify.app/docs/options/#modifierkey
  const delta = deltaY * (evt.ctrlKey ? 3 : 1);
  const zoom = getBoundedValue(canvas.getZoom() * 0.999 ** delta, 0.01, 20);
  canvas.zoomToPoint({ x, y }, zoom);
};

const setViewport = (x: number = 0, y: number = 0) => {
  const zoom = canvas.getZoom();
  const [w, h] = [canvas.width!, canvas.height!];
  const vpt = canvas.viewportTransform as number[];
  const offsetX = w - canvasInitSize!.w * zoom;
  const offsetY = h - canvasInitSize!.h * zoom;
  vpt[4] =
    offsetX > 0 ? offsetX / 2 : getBoundedValue(vpt[4] + x, offsetX, 0);
  vpt[5] =
    offsetY > 0 ? offsetY / 2 : getBoundedValue(vpt[5] + y, offsetY, 0);
  canvas.requestRenderAll();
  canvas.getObjects().forEach((obj) => obj.setCoords());
};

const listeners = {
  'mouse:wheel': (e: fabric.IEvent<Event>) => {
      const { evt } = parseEvent(e as fabric.IEvent<WheelEvent>);

      setZoomByWheel(evt);
      setViewport();
    },
  };

  'mouse:down': (e: fabric.IEvent<Event>) => {
    const { pointer } = parseEvent(e as fabric.IEvent<MouseEvent>);

    isPanning.current = true;
    lastPosition.current = pointer!;
    canvas.setCursor('grabbing');
  },

  'mouse:move': (e: fabric.IEvent<Event>) => {
    if (!isPanning.current) return;

    const { pointer } = parseEvent(e as fabric.IEvent<MouseEvent>);
    setViewport(
      pointer!.x - lastPosition.current.x,
      pointer!.y - lastPosition.current.y
    );
    canvas.setViewportTransform(canvas.viewportTransform as number[]);
    lastPosition.current = pointer!;
    canvas.setCursor('grabbing');
  },

  'mouse:up': (e: fabric.IEvent<Event>) => {
    isPanning.current = false;
    canvas.setCursor('default');
  },
};
```

In v6, i found if using the same code, the zooming is fine, but panning is very small feels lagging after canvas is zoomed in. I noticed `pointer` and `canvas.getPointer()` always return same value, but logics in `setViewPort()` is expecting absolute coords, e.g. w, h. I noticed in v6 `canvas.getPointer(e: TPointerEvent, ignoreVpt?: boolean | undefined)` `ignoreVpt` defaults to `false` giving canvas coords, in `canvas.fire('mouse:over', opts)` `ops`, there is an `absolutePointer` prop required.  So I wondered if I can get `absolutePointer` from v6's events, and I tried and it worked, `absolutePointer` gives different result from `pointer`/`canvas.getPointer()`, now code looks like this:

```typescript
import { TPointerEventInfo as IEvent } from 'fabric/src/EventTypeDefs';

function parseEvent<T extends MouseEvent | WheelEvent>(e: IEvent<T>) {
  const { button, target, pointer, e: evt, absolutePointer } = e;
  evt.preventDefault();
  evt.stopPropagation();

  return { button, target, pointer, evt, absolutePointer };
}

const listeners = {
  'mouse:down': (e: IEvent<MouseEvent>) => {
    const { absolutePointer } = parseEvent(e);

    isPanning.current = true;
    lastPosition.current = absolutePointer;
    canvas.setCursor('grabbing');
    selectLabels([]);
  },

  'mouse:move': (e: IEvent<MouseEvent>) => {
    if (!isPanning.current) return;

    const { absolutePointer } = parseEvent(e as fabric.IEvent<MouseEvent>);
    setViewport(
      absolutePointer.x - lastPosition.current.x,
      absolutePointer.y - lastPosition.current.y
    );
    canvas.setViewportTransform(canvas.viewportTransform);
    lastPosition.current = absolutePointer;
    canvas.setCursor('grabbing');
  },

  // other listeners as before
}
```

8. During my debugging yesterday for this issue, I found the following trace that fires the extra `object:modified`
- https://github.com/fabricjs/fabric.js/blob/196bea176c4b01fd68a05aa096ba5d22459201b8/src/canvas/StaticCanvas.ts#L306 , `super().remove()` goes to 
-  https://github.com/fabricjs/fabric.js/blob/196bea176c4b01fd68a05aa096ba5d22459201b8/src/Collection.ts#L59, `this._onObjectRemoved()` goes to
- https://github.com/fabricjs/fabric.js/blob/196bea176c4b01fd68a05aa096ba5d22459201b8/src/canvas/SelectableCanvas.ts#L528, noticed here it checks if is `active`, `this._discardActiveObject()` goes to
- https://github.com/fabricjs/fabric.js/blob/196bea176c4b01fd68a05aa096ba5d22459201b8/src/canvas/SelectableCanvas.ts#L1427, then goes to
- https://github.com/fabricjs/fabric.js/blob/196bea176c4b01fd68a05aa096ba5d22459201b8/src/canvas/Canvas.ts#L960, then goes to
- https://github.com/fabricjs/fabric.js/blob/196bea176c4b01fd68a05aa096ba5d22459201b8/src/canvas/Canvas.ts#L988-L991

**Reply by ShaMan123** [reply]
For 8 I would like a live reproduction. I want to take a look

**Reply by ZhengRui** [reply]
@ShaMan123 , you are right, normally delete objects won't trigger `object:modified` event. My case is an edge case that i delete all the objects inside `object:modified` callback, which triggers `object:modified` event again because at that moment the active object's `transform.actionPerformed` is not null, so it fired an unexpected `object:modified` event as shown 👆

**Reply by ShaMan123** [reply]
an edge cases indeed. Would love seeing a live demo of it

**Reply by ZhengRui** [reply]
[demo](https://fabric-repro.vercel.app/objectmodified), [code](https://github.com/ZhengRui/fabric-repro/blob/main/src/app/objectmodified/page.tsx)

**Reply by ShaMan123** [reply]
please try #8752 that should fix this issue

**Comment by ZhengRui** [comment]
the last issue i am having now is animation, @ShaMan123 can you give some guidence how to adapt this part to v6? Thank you.

```typescript
    if (startOrStop === 'start')
      animation.current = setInterval(
        () =>
          circle.animate(
            'radius',
            circle.radius === RADIUS ? 2.0 * RADIUS : RADIUS,
            {
              duration: 200,
              onChange: canvas.renderAll.bind(canvas),
              easing: fabric.util.ease.easeInOutCubic,
            }
          ),
        400
      );

    if (startOrStop === 'stop') {
      clearInterval(animation.current!);
      fabric.runningAnimations.cancelAll();
      circle.set({
        radius: selectedPids.includes(pid) ? 1.5 * RADIUS : RADIUS,
      });
      canvas.requestRenderAll();
    }

```

**Reply by ShaMan123** [reply]
https://github.com/fabricjs/fabric.js/issues/8299#issuecomment-1447265061

```diff
    if (startOrStop === 'start')
      animation.current = setInterval(
        () =>
-          circle.animate(
+          circle._animate(
            'radius',
            circle.radius === RADIUS ? 2.0 * RADIUS : RADIUS,
            {
              duration: 200,
              onChange: canvas.renderAll.bind(canvas),
              easing: fabric.util.ease.easeInOutCubic,
            }
          ),
        400
      );

    if (startOrStop === 'stop') {
      clearInterval(animation.current!);
      fabric.runningAnimations.cancelAll();
      circle.set({
        radius: selectedPids.includes(pid) ? 1.5 * RADIUS : RADIUS,
      });
      canvas.requestRenderAll();
    }
```

I would also use the `delay` options instead of wrapping with `setInterval`

```diff
    if (startOrStop === 'start')
      animation.current = () =>
          circle._animate(
            'radius',
            circle.radius === RADIUS ? 2.0 * RADIUS : RADIUS,
            {
              duration: 200,
              onChange: () => canvas.renderAll(),
              easing: fabric.util.ease.easeInOutCubic,
+              delay: 400,
+             onComplete: () => callSelf()
            }
          );

    if (startOrStop === 'stop') {
+      animation.current.cancel();
-      fabric.runningAnimations.cancelAll();
      circle.set({
        radius: selectedPids.includes(pid) ? 1.5 * RADIUS : RADIUS,
      });
      canvas.requestRenderAll();
    }
```

**Reply by ZhengRui** [reply]
Thanks @ShaMan123. Two things got the animation working:
1. `circle._animate()`
2. this is weird thing, I have to manually/explicitly `setRadius` inside `onChange` to get it working. Without this hack, the radius props did change when `console.log`, but not reflected visually in the canvas (even with `canvas.renderAll()` / `canvas.requestRenderAll()`). 
```typescript
onChange: () => {
  circle.setRadius(circle.radius);
  canvas.requestRenderAll();
},
```

**Reply by ShaMan123** [reply]
this doesn't make sense... I would like a reproduction of this as well please

**Reply by ZhengRui** [reply]
@ShaMan123  you can see the demo here: https://fabric-repro.vercel.app/animation, code is here https://github.com/ZhengRui/fabric-repro/blob/main/src/app/animation/page.tsx , you can see circle radius always change correctly, but without the setRadius hack, visually the animation doesn't show up.

**Reply by ShaMan123** [reply]
Now I understand.
Animation doesn't change anything. It just calculates the values for you and exposes them in callbacks. You are in charge of the side effect. It is how it is built since side effects can vary a lot.

**Comment by iocoker** [comment]
Hi guys, how do we define new objects on fabric?
In old system, there's:
`class Calendar extends fabric.Group {....}`
`fabric.util.classRegistry.setClass(Calendar);`
`fabric.Calendar = Calendar`
And then you can do: `new fabric.Calendar()`

But `import * as fabric from 'fabric'` makes `fabric.Calendar = Calendar` impossible.
Any thoughts?

**Reply by ShaMan123** [reply]
the whole point of the registry is for this.
exports are frozen, you can't mutate them.
Once you registered it fabric is good to go. import your class and use it and fabric will query the registry when it needs to, loading from json for example

**Reply by iocoker** [reply]
I see. Thanks!

**Reply by iocoker** [reply]
So, I made it generic class instead of attaching it to fabric, then registered it and worked fine.

But it was necessary to add the second parameter `fabric.util.classRegistry.setClass(Calendar, 'calendar');`

**Reply by ShaMan123** [reply]
We are working on that right now #8714

**Reply by ShaMan123** [reply]
notice that master has broken beta
`fabric.util.classRegistry` => `fabric.classRegistry`

**Comment by ghost** [comment]
Haven't been able to confirm but looks like in v6:

```
line.x2 = something; // is different than
line.set("x2", something);
```
